File: Self_Dividing_Numbers_/py.d/self_divide.py
from typing import List

class Solution:
  def selfDividingNumbers(self, left: int, right: int) -> List[int]:
    zero_mods = []
    for i in range(left, right + 1):
      mod_flag = True
      for num in [int(j) for j in str(i)]:
        if num == 0 or i % num != 0:
          mod_flag = False
      if mod_flag:
        zero_mods.append(i)
    return zero_mods

if __name__=="__main__":
  sol = Solution()
  print(sol.selfDividingNumbers(1, 22)) # => [1,2,3,4,5,6,7,8,9,11,12,15,22]

File: Self_Dividing_Numbers/cpp.d/self_divide.cpp
#include <iostream>
#include <vector>
#include <stdio.h>
using namespace std;


class Solution {
private:
  bool div_zero_flag;
  void divisible_int(int sub_num, int num){
    if (sub_num >= 10 and this->div_zero_flag)
      divisible_int(sub_num / 10, num);
    sub_num = sub_num % 10;
    cout << "Testing divisibility of " << num << "/" << sub_num << endl;
    if (sub_num != 0 && num % sub_num == 0){
      cout << num << " is divisible by " << sub_num << endl;
    } else {
      cout << num << " is not divisible by " << sub_num << "!!!" << endl;
      this->div_zero_flag = false;
    }
    cout << endl;
  }

public:
  vector<int> selfDividingNumbers(int left, int right){
    vector<int> self_divide;
    for (;left <= right; left++){
      this->div_zero_flag = true;
      this->divisible_int(left, left);
      if (this->div_zero_flag){
        cout << left << " is self divisible; appending" << endl;
        self_divide.push_back(left);
      }
      cout << "Done iterating for " << left << endl << endl;
    }

    return self_divide;
  }
};


int main(){
  int left = 1, right = 22; // => => [1,2,3,4,5,6,7,8,9,11,12,15,22]
  Solution sol;
  vector<int> nums = sol.selfDividingNumbers(left, right);
  for (auto x: nums)
    cout << x << " ";
  cout << endl;
  return 0;
}
